k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java

```diff
@@ -0,0 +1,409 @@
+package com.fsck.k9.mail.store.imap.command;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+import android.text.TextUtils;
+
+import com.fsck.k9.mail.Flag;
+import com.fsck.k9.mail.MessageRetrievalListener;
+import com.fsck.k9.mail.MessagingException;
+import com.fsck.k9.mail.store.imap.ImapFolder;
+import com.fsck.k9.mail.store.imap.ImapMessage;
+import com.fsck.k9.mail.store.imap.ImapResponse;
+import com.fsck.k9.mail.store.imap.ImapUtility;
+import com.fsck.k9.mail.store.imap.response.SearchResponse;
+
+public class UidSearchCommand extends BaseCommand {
+
+    private ImapFolder folder;
+    private MessageRetrievalListener<ImapMessage> listener;
+
+    private Set<Long> sequenceSet;
+    private List<Map.Entry<String, String>> sequenceRanges;
```

Comment:
I was thinking of making this into `List<Map.Entry<Long, Long>>`  to make it easier for comparison while minimizing the number of elements.